### PR TITLE
useBaseField: Convert component to TypeScript

### PR DESCRIPTION
**Pull request by walbo**
## What?
Refactor the `useBaseField` component to TypeScript.

## Why?
Part of the @wordpress/components's TypeScript migration (https://github.com/WordPress/gutenberg/issues/35744).

## How?
- Refactored the component TypeScript
- Refactored tests to TypeScript

## Testing Instructions
1. Run `npm run dev` and ensure no typing errors
2. Run  `npm run test:unit -- packages/components/src/base-field` and make sure unit tests passes.

### CODE REVIEW

File: packages/components/src/base-field/types.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/**
+ * External dependencies
+ */
+import type { ReactNode } from 'react';
+
+export type BaseFieldProps = {
```
Comment (ciampo) [inline packages/components/src/base-field/types.ts]: Looking at the typedefs in the old JavaScript file, I think we're missing `FlexProps`

Comment (walbo) [reply packages/components/src/base-field/types.ts]: Ups.

I'll update the PR later today.

File: packages/components/src/base-field/types.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	 */
+	hasError?: boolean;
+	/**
+	 * Renders a component that can be inlined in some text.
+	 */
+	isInline?: boolean;
+	/**
+	 * Renders a subtle variant of the component.
+	 */
+	isSubtle?: boolean;
```
Comment (ciampo) [inline packages/components/src/base-field/types.ts]: I believe these are missing their `@default` values in the JSDocs (and in the README)

File: packages/components/src/base-field/test/index.tsx [hunk 3]
```diff
@@ -8,9 +8,13 @@ import { render, screen } from '@testing-library/react';
  */
 import { useBaseField } from '../index';
 import { View } from '../../view';
+import type { WordPressComponentProps } from '../../ui/context';
+import type { BaseFieldProps } from '../types';
 
-const TestField = ( props ) => {
-	return <View { ...useBaseField( props ) } />;
+const TestField = (
+	props: Omit< WordPressComponentProps< BaseFieldProps, 'div' >, 'children' >
```
Comment (ciampo) [inline packages/components/src/base-field/test/index.tsx]: Nit: I don't think it's necessary to use `WordPressComponentProps` in this test, `Omit< BaseFieldProps, 'children' >` should suffice

**Review by ciampo — COMMENTED** [review]
Thank you for iterating on this! I gave it another look and left a couple more comments (sorry for not noticing before!)

**Comment by ciampo** [comment]
Hey @walbo , let me know if you don't have capacity to work on this, and I'll be happily take over :)

**Comment by walbo** [comment]
Hi @ciampo 

Yes, I'm a bit busy with other commitments the next couple of weeks so if you don’t mind I would appreciate if you could land this one 👍 Thanks

**Comment by ciampo** [comment]
Absolutely, no issues! Thank you anyway for starting the work on this one!

File: packages/components/src/base-field/types.ts (hunk 2 above)
Comment (ciampo) [reply packages/components/src/base-field/types.ts]: Done in https://github.com/WordPress/gutenberg/pull/45712/commits/559e5eeaeb37edea92fc7638506f6328d6895db6

File: packages/components/src/base-field/test/index.tsx (hunk 3 above)
Comment (ciampo) [reply packages/components/src/base-field/test/index.tsx]: Done in https://github.com/WordPress/gutenberg/pull/45712/commits/dc1ffdc917712a3c405e5fc6d5b5ad15c330dd89

**Review by ciampo — APPROVED** [review]

**Pull request merged**